csharp/src/Microsoft.ML.OnnxRuntime/OrtIoBinding.cs

```diff
@@ -0,0 +1,245 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Runtime.InteropServices;
+
+namespace Microsoft.ML.OnnxRuntime
+{
+    /// <summary>
+    /// This class enable to bind inputs and outputs to pre-allocated
+    /// memory. This enables interesting scenarios. For example, if your input
+    /// already resides in some pre-allocated memory even if on a device you bind
+    /// that piece of memory to an input name and shape and onnxruntime will use that as input.
+    /// Other traditional inputs can also be bound that already exists as Tensors
+    /// </summary>
+    public class OrtIoBinding : IDisposable
+    {
+        private IntPtr _handle;
+
+        /// <summary>
+        /// Use InferenceSession.CreateIoBinding()
+        /// </summary>
+        /// <param name="session"></param>
+        internal OrtIoBinding(InferenceSession session)
+        {
+            NativeApiStatus.VerifySuccess(NativeMethods.OrtCreateIoBinding(session.Handle, out _handle));
+        }
+
+        internal IntPtr Handle
+        {
+            get
+            {
+                return _handle;
+            }
+        }
+
+        /// <summary>
+        /// Bind a piece of pre-allocated native memory as a OrtValue Tensor with a given shape
+        /// to an input with a given name. The model will read the specified input from that memory
+        /// possibly avoiding the need to copy between devices. OrtMemoryAllocation continues to own
+        /// the chunk of native memory and should be alive until the end of execution.
+        /// The size of the allocation can not be less than required.
+        /// by the Tensor of the given size.
+        /// </summary>
+        /// <param name="name">of the input</param>
+        /// <param name="elementType">Tensor element type</param>
+        /// <param name="shape"></param>
+        /// <param name="allocation">native memory allocation</param>
+        public void BindInput(string name, Tensors.TensorElementType elementType, long[] shape, OrtMemoryAllocation allocation)
+        {
+            using (var ortValue = OrtValue.CreateTensorValueWithData(allocation.Info,
+                                                                    elementType,
+                                                                    shape,
+                                                                    allocation.Pointer, allocation.Size))
+                BindIntputOrOutput(name, ortValue.Handle, true);
+        }
+
+        /// <summary>
+        /// Bind the input with the given name as an OrtValue Tensor allocated in pinned managed memory.
+        /// Instance of FixedBufferOnnxValue owns the memory and should be alive until the end of execution.
+        /// </summary>
+        /// <param name="name">name of input</param>
+        /// <param name="fixedValue"></param>
+        public void BindInput(string name, FixedBufferOnnxValue fixedValue)
```

Comment:
From what I understand, you provide two ways to create bound inputs. 1. Using OrtMemoryAllocation with a shape 2. With FixedBufferOnnxValue. 
OrtMemoryAllocation, does not allow you to read or write to the allocated memory. So, in the case when I want to bind an input with some initial values, I would need to use FixedBufferOnnxValue. Is my interpretation right? 
Also, in that case the underlying memory buffer would be the exact same size as the input tensor. But in case of OrtMemoryAllocation, the allocated buffer size can be larger than the shape used when binding the input. 

If my interpretation is correct, this would work for our case. But it would be nice to have a read and write method the OrtMemoryAllocation objects.

Reply:
I think your understanding is spot on. You can actually read and write native memory OrtMemoyrAllocation using unsafe code and there is an example in the test. You can't, however, do so with CUDA on device memory. We have debating on whether to provide it and decided to do so at a later date if needed. The reason is bc as you mentioned, your use case for now is satisfied.
Another potential way to read/write CUDA memory is to allocate some sort of host accessible CUDA memory. I have seen a mention of it, but I am not sure if I understood it correctly. If true, this mechanism would enable allocating such memory.